File: worker/src/RTC/RtpObserver.cpp
#define MS_CLASS "RTC::RtpObserver"
// #define MS_LOG_DEV_LEVEL 3

#include "RTC/RtpObserver.hpp"
#include "Logger.hpp"
#include "MediaSoupErrors.hpp"

namespace RTC
{
	/* Instance methods. */

	RtpObserver::RtpObserver(RTC::Shared* shared, const std::string& id, RTC::RtpObserver::Listener* listener)
	  : id(id), shared(shared), listener(listener)
	{
		MS_TRACE();
	}

	RtpObserver::~RtpObserver()
	{
		MS_TRACE();
	}

	void RtpObserver::HandleRequest(Channel::ChannelRequest* request)
	{
		MS_TRACE();

		switch (request->method)
		{
			case Channel::ChannelRequest::Method::RTPOBSERVER_PAUSE:
			{
				this->Pause();

				request->Accept();

				break;
			}

			case Channel::ChannelRequest::Method::RTPOBSERVER_RESUME:
			{
				this->Resume();

				request->Accept();

				break;
			}

			case Channel::ChannelRequest::Method::RTPOBSERVER_ADD_PRODUCER:
			{
				const auto* body = request->data->body_as<FBS::RtpObserver::AddProducerRequest>();
				auto producerId  = body->producerId()->str();

				RTC::Producer* producer = this->listener->RtpObserverGetProducer(this, producerId);

				this->AddProducer(producer);

				this->listener->OnRtpObserverAddProducer(this, producer);

				request->Accept();

				break;
			}

			case Channel::ChannelRequest::Method::RTPOBSERVER_REMOVE_PRODUCER:
			{
				const auto* body = request->data->body_as<FBS::RtpObserver::RemoveProducerRequest>();
				auto producerId  = body->producerId()->str();

				RTC::Producer* producer = this->listener->RtpObserverGetProducer(this, producerId);

				this->RemoveProducer(producer);

				// Remove from the map.
				this->listener->OnRtpObserverRemoveProducer(this, producer);

				request->Accept();

				break;
			}

			default:
			{
				MS_THROW_ERROR("unknown method '%s'", request->methodCStr);
			}
		}
	}

	void RtpObserver::Pause()
	{
		MS_TRACE();

		if (this->paused)
		{
			return;
		}

		this->paused = true;

		Paused();
	}

	void RtpObserver::Resume()
	{
		MS_TRACE();

		if (!this->paused)
		{
			return;
		}

		this->paused = false;

		Resumed();
	}
} // namespace RTC

File: node/src/RtpObserver.ts
import { Logger } from './Logger';
import { EnhancedEventEmitter } from './enhancedEvents';
import type {
	RtpObserverEvents,
	RtpObserverObserver,
} from './RtpObserverTypes';
import type { Channel } from './Channel';
import type { RouterInternal } from './Router';
import type { Producer } from './ProducerTypes';
import type { AppData } from './types';
import * as FbsRequest from './fbs/request';
import * as FbsRouter from './fbs/router';
import * as FbsRtpObserver from './fbs/rtp-observer';

export type RtpObserverConstructorOptions<RtpObserverAppData> = {
	internal: RtpObserverObserverInternal;
	channel: Channel;
	appData?: RtpObserverAppData;
	getProducerById: (producerId: string) => Producer | undefined;
};

type RtpObserverObserverInternal = RouterInternal & {
	rtpObserverId: string;
};

const logger = new Logger('RtpObserver');

export abstract class RtpObserverImpl<
	RtpObserverAppData extends AppData = AppData,
	Events extends RtpObserverEvents = RtpObserverEvents,
	Observer extends RtpObserverObserver = RtpObserverObserver,
> extends EnhancedEventEmitter<Events> {
	// Internal data.
	protected readonly internal: RtpObserverObserverInternal;

	// Channel instance.
	protected readonly channel: Channel;

	// Closed flag.
	#closed = false;

	// Paused flag.
	#paused = false;

	// Custom app data.
	#appData: RtpObserverAppData;

	// Method to retrieve a Producer.
	protected readonly getProducerById: (
		producerId: string
	) => Producer | undefined;

	// Observer instance.
	readonly #observer: Observer;

	protected constructor(
		{
			internal,
			channel,
			appData,
			getProducerById,
		}: RtpObserverConstructorOptions<RtpObserverAppData>,
		observer: Observer
	) {
		super();

		logger.debug('constructor()');

		this.internal = internal;
		this.channel = channel;
		this.#appData = appData ?? ({} as RtpObserverAppData);
		this.getProducerById = getProducerById;
		this.#observer = observer;
	}

	get id(): string {
		return this.internal.rtpObserverId;
	}

	get closed(): boolean {
		return this.#closed;
	}

	get paused(): boolean {
		return this.#paused;
	}

	get appData(): RtpObserverAppData {
		return this.#appData;
	}

	set appData(appData: RtpObserverAppData) {
		this.#appData = appData;
	}

	get observer(): Observer {
		return this.#observer;
	}

	close(): void {
		if (this.#closed) {
			return;
		}

		logger.debug('close()');

		this.#closed = true;

		// Remove notification subscriptions.
		this.channel.removeAllListeners(this.internal.rtpObserverId);

		/* Build Request. */
		const requestOffset = new FbsRouter.CloseRtpObserverRequestT(
			this.internal.rtpObserverId
		).pack(this.channel.bufferBuilder);

		this.channel
			.request(
				FbsRequest.Method.ROUTER_CLOSE_RTPOBSERVER,
				FbsRequest.Body.Router_CloseRtpObserverRequest,
				requestOffset,
				this.internal.routerId
			)
			.catch(() => {});

		this.emit('@close');

		// Emit observer event.
		this.#observer.safeEmit('close');
	}

	routerClosed(): void {
		if (this.#closed) {
			return;
		}

		logger.debug('routerClosed()');

		this.#closed = true;

		// Remove notification subscriptions.
		this.channel.removeAllListeners(this.internal.rtpObserverId);

		this.safeEmit('routerclose');

		// Emit observer event.
		this.#observer.safeEmit('close');
	}

	async pause(): Promise<void> {
		logger.debug('pause()');

		const wasPaused = this.#paused;

		await this.channel.request(
			FbsRequest.Method.RTPOBSERVER_PAUSE,
			undefined,
			undefined,
			this.internal.rtpObserverId
		);

		this.#paused = true;

		// Emit observer event.
		if (!wasPaused) {
			this.#observer.safeEmit('pause');
		}
	}

	async resume(): Promise<void> {
		logger.debug('resume()');

		const wasPaused = this.#paused;

		await this.channel.request(
			FbsRequest.Method.RTPOBSERVER_RESUME,
			undefined,
			undefined,
			this.internal.rtpObserverId
		);

		this.#paused = false;

		// Emit observer event.
		if (wasPaused) {
			this.#observer.safeEmit('resume');
		}
	}

	async addProducer({ producerId }: { producerId: string }): Promise<void> {
		logger.debug('addProducer()');

		const producer = this.getProducerById(producerId);

		if (!producer) {
			throw Error(`Producer with id "${producerId}" not found`);
		}

		const requestOffset = new FbsRtpObserver.AddProducerRequestT(
			producerId
		).pack(this.channel.bufferBuilder);

		await this.channel.request(
			FbsRequest.Method.RTPOBSERVER_ADD_PRODUCER,
			FbsRequest.Body.RtpObserver_AddProducerRequest,
			requestOffset,
			this.internal.rtpObserverId
		);

		// Emit observer event.
		this.#observer.safeEmit('addproducer', producer);
	}

	async removeProducer({ producerId }: { producerId: string }): Promise<void> {
		logger.debug('removeProducer()');

		const producer = this.getProducerById(producerId);

		if (!producer) {
			throw Error(`Producer with id "${producerId}" not found`);
		}

		const requestOffset = new FbsRtpObserver.RemoveProducerRequestT(
			producerId
		).pack(this.channel.bufferBuilder);

		await this.channel.request(
			FbsRequest.Method.RTPOBSERVER_REMOVE_PRODUCER,
			FbsRequest.Body.RtpObserver_RemoveProducerRequest,
			requestOffset,
			this.internal.rtpObserverId
		);

		// Emit observer event.
		this.#observer.safeEmit('removeproducer', producer);
	}
}
